lib/netext/httpext/transport.go

```diff
@@ -80,68 +80,64 @@ func (t *transport) measureAndEmitMetrics(unfReq *unfinishedRequest) *finishedRe
 		trail:             trail,
 	}
 
-	tags := t.tags
+	ctm := t.tagsAndMeta.Clone()
 	enabledTags := t.state.Options.SystemTags
-	urlEnabled := enabledTags.Has(metrics.TagURL)
-	var setName bool
-	if _, ok := tags.Get("name"); !ok && enabledTags.Has(metrics.TagName) {
-		setName = true
-	}
-	if urlEnabled || setName {
-		cleanURL := URL{u: unfReq.request.URL, URL: unfReq.request.URL.String()}.Clean()
-		if urlEnabled {
-			tags = tags.With("url", cleanURL)
-		}
-		if setName {
-			tags = tags.With("name", cleanURL)
+	cleanURL := URL{u: unfReq.request.URL, URL: unfReq.request.URL.String()}.Clean()
+
+	// After k6 v0.41.0, the `name` and `url` tags  have the exact same values:
+	nameTagValue, nameTagManuallySet := ctm.Tags.Get(metrics.TagName.String())
+	if !nameTagManuallySet {
+		// If the user *didn't* manually set a `name` tag value and didn't use
+		// the http.url`` template literal helper to have k6 automatically set
+		// it (see `lib/netext/httpext.MakeRequest()`), we will use the cleaned
+		// URL value as the value of both `name` and `url` tags.
+		ctm.SetSystemTagOrMetaIfEnabled(enabledTags, metrics.TagName, cleanURL)
+		ctm.SetSystemTagOrMetaIfEnabled(enabledTags, metrics.TagURL, cleanURL)
+	} else {
+		// However, if the user set the `name` tag value somehow, we will use
+		// whatever they set as the value of the `url` tags too, to prevent
+		// high-cardinality values in the indexed tags.
+		ctm.SetSystemTagOrMetaIfEnabled(enabledTags, metrics.TagURL, nameTagValue)
+
+		// However, if the actual clean-URL value would have been different than
+		// the `name` value, we will also attach it as the `raw_url` non-indexed
+		// tag (i.e. add it to the Sample.Metadata).
+		if nameTagValue != cleanURL {
+			ctm.SetSystemTagOrMetaIfEnabled(enabledTags, metrics.TagRawURL, cleanURL)
```

Comment:
:+1: for points 2. and 3., I will fix them.

I am not so sure about point 1. though :thinking: We don't always set all possible tags on all metric samples. For example, we only set the `error` and `error_code` tags if there was actually an error with the request. So I think it makes some sense to only set the `raw_url` tag when the actual URL of the HTTP request is going to be different from the value of the `url` tag, i.e. when there is URL grouping with `name`. 

> which means that it will be hard to use for the people who might want to use it directly

Why? Users can't set a threshold on it, and `Response.url` will be the actual URL. So it's mostly going to be exposed for things like the `json` and `csv` outputs and potentially for exporting tracing data, where it doesn't seem like a problem. Besides, a user can manually set the `raw_url` tag on requests that won't automatically have it assigned, if they really want to.

So, I am leaning towards :-1: for point 1. :thinking: @grafana/k6-devs, what do you think?

Reply:
An alternative suggestion. Could it be an option to not include `raw_url` in the core? The user has the option to add it manually so we could apply it in this way for now. The tracing library could do the same. We could provide a js helper in utils if it's really required. WDYT?

Reply:
hmm, implementing this purely in JS is going to be tricky, e.g. if you have a `name` tag to some request that has redirects :thinking:

Reply:
I am currently for dropping it from this PR and add it later. 

I am still of the opinion that this likely will not be used by 99% of the users. 

And even less so for the people who want to name their redirects. 

WDYT? dropp `raw_url` open PR for later?

Reply:
to get back to your original points 2. and 3. - I tried to fix them, but the resulting code is a bit nasty, and doesn't really optimize much. Right now, this allocation:
```go
cleanURL := URL{u: unfReq.request.URL, URL: unfReq.request.URL.String()}.Clean()
```

just allocates the `URL` object directly on the stack, and that URL object is very tiny: https://github.com/grafana/k6/blob/7c90538db8b799df11022500cb81d91c4e223dc1/lib/netext/httpext/url.go#L9-L14

And, if the actual URL doesn't have credentials in it, the `URL.Clean()` method doesn't allocate anything new, it just returns one of the existing object properties: https://github.com/grafana/k6/blob/7c90538db8b799df11022500cb81d91c4e223dc1/lib/netext/httpext/url.go#L34-L40 

So, we don't gain anything significant and the cost is a more difficult to follow code, so I'd prefer to leave points 2. and 3. as they are too.

Reply:
> if we drop raw_url, then the breaking change we are doing to url will be much bigger IMO.

welll .. I guess we disagree on this as well as on other things.

Irregardless of whether `raw_url` is present or not, the `url` breaking change is a breaking change. To fix it users will need to:
1. care about this - which is the other part that I think you are *extremely* overestimating how many users (if any) care. We have never had this as funcitionality in the cloud. I don't remember this being a problem. Not to speak that any system that needs to intake this data will need to be able to ingest it - which is not really possible with any of the current outputs for most RPSs.
2. They will still need to actually change to using `raw_url` instead of `url`. Which might be a lot more involved.
3. In this PR all but 2 outputs - csv and json, does *not* support outputing the metadata. So it still "much bigger" breaking change for them.

Given this, for me `raw_url` is fixing the problem half-arsedly for some percentage of all users and for users that IMO just don't exist. 

If users want to know their original `url` they likely won't be using `name` in the first place ;). 

Also with the current imlpementation of `1` it means that some requests will have to be queried (for simplicity in grphana) for `url` and some for `raw_url` to get their url. 

But again IMO this helps noone and not in a way that will even matter without even *more* changes.

So I propose we drop it and discuss it after the release if we even should provide it for users. Or wait for users to come and ask about it.

Reply:
Yeah, I guess we disagree :slightly_smiling_face: In my mind, it is very simple, and `raw_url` satisfies a very important use case with very little cost. 

With the current k6 version, if you have some outlier request that times out or just takes too long, you can find its actual URL, even if you've specified a `name` tag. Yes, you'd need to also export your granular non-aggregated results to either a .json, or a .csv output, or stream them to InfluxDB, but you can do it. 

This is the capability I am trying to preserve with `raw_url`. And, if we don't have `raw_url`, users will literally have no way to diagnose such an issue, unless they write a wrapper around `k6/http` that manually looks at `Response.timings` and logs something... :confused: 

I haven't added support for non-indexed tags as fields in the `influxdb` output just because that output is basically on life-support... But if that will help you accept this PR, I can easily add that too.

Having `raw_url` will also be helpful in other things. For example, once we have tracing or some output with exemplars and/or sampling of metric outliers, we should probably include the `raw_url` (and other non-indexed tags) there too, since they may help tremendously with the debugging and we won't have concerns about their high cardinality.

Reply:
So your argument is that we need this because if someone: 
1. uses `name` tag
2. has a particular request with a particular url that is hidden by `name` but is very problematic - let's say takes 10x longer or returns some error. 
3. has an output that supports this *and* the capacity to actually ingest the data. 

We need to have a separate non indexable tags that will be on by default for everybody? 

IMO in this particular case the user can just use the workaround @codebien provided. They still need to do extra work to figure this out - so why should we make everybody have `raw_url`.


I am still fo the opinion that my original `1` is very strange UX decision

Reply:
> so why should we make everybody have raw_url.

we don't make anyone have anything :smile: I want us to provide the `raw_url` value by default, which might be of use to a subset of users that are diagnosing an issue with their systems :wink: and I'd argue it has a very negligible cost

> I am still fo the opinion that my original 1 is very strange UX decision

should we always send the `error` and `error_code` tags then, even on successful requests? :thinking: or the TLS version for non-HTTP requests? :sweat_smile: there are a lot of precedents of k6 not sending all possible tags for every metric, and I don't see us changing that

Reply:
> I'd argue it has a very negligible cos

We will always allocate a map and a string even if most users never use them (or have a way to use them). And this happens specifically in the case where the majority of the users of `name` will be doing it in order to get *better* performance, and because the concrete url is not important ;). 

Again IMO the users you are some very small minority and there is a workaround for their use case. 

We can also always add this *later* if I am wrong and there is a big majority of users needing it. But as time has shown we rarely if ever remove stuff.

> should we always send the error and error_code tags then, even on successful requests? thinking or the TLS version for non-HTTP requests? sweat_smile there are a lot of precedents of k6 not sending all possible tags for every metric, and I don't see us changing that

No - an empty `error` when you are querying for an `error` will tell you that there is no error. The same as having an *empty* error. Same for `error_code` or any of the other suggestions.

Querying for `raw_url` and getting empty doesn't mean that hte `url` is empty - it means the `url` is the same as the `name`.  This will now need to be documented and if somebody doesn't read it and just sees "ooh tehre is an raw_url in my data" and uses it - can very easily later miss all the requests that do *not* have `raw_url`. 

Arguably all of the above mean that this needs to be discussed. And I would prefer to do it outside of this PR instead of blocking the PR on this particular issue. So this is why I propose to just drop it and have the discussion outside of the PR. Might even happen within this release :shrug: